### ISSUE TITLE
Can you help with custom app?

### ISSUE THREAD

**Issue by MrBenFTW** [relikd/AppCheck#1]
Hi, I am trying to create an app which is literally just a connect button which routes all DNS traffic to my custom DNS server which blocks certain sites and services. All the apps I found on the App Store are a bit too complex. I basically want a pseudo vpn server running on the device that only does DNS queries, rather than actually forwarding traffic outside of the device and changing the IP etc, it only sends the dns. Effectively acting as if you change the dns settings in wifi, except it is always on demand and works on cellular data too.

In iOS 14 apple added this as a native ability I know, however, I need it to work with iOS 13 and maybe 12 as well. Essentially, DNS cloak but with a hardcoded dns IP inside of the app and a single view with just a connect button.

I know how to make the app myself, UI etc, I can do swift. But when it comes to Network Extensions I am lost. If you feel like you could make a simple project which just has something like let dnsServer = ["1.1.1.1", "1.0.0.1"] then I can pay you for it. Then I can make the rest of the app myself

**Comment by relikd** [comment]
What is your requirement exactly? Do you need to send the data to your DNS server or can this also be handled by the app solely?  
E.g. does your server contain complex logic or is it a simple domain list, that you want to block? In the latter case the app could contain that list and block outgoing traffic without ever contacting your server. This does not even need to be a static list. You could have an additional sync button inside your app that will update the blocking list from your server.

**Comment by relikd** [comment]
Or do you want just a plain DNS server. Which wouldn't even require to use NEKit as a dependency. Either way, I did both for you 😄. 
PS. I know how hard it is to get around with NetworkExtensions. Took me a few weeks to figure that out.

**Comment by MrBenFTW** [comment]
Just a plain dns client to connect to my DNS server. Simple connect button which connects to it as a 'vpn' from a predefined ip something like let dnsIP=

and then a simple method to call connectDNS() and disconnectDNS()

I have used NEKit before but it seems like its been depreciated now and it was way to complex for me. I would rather something standalone no dependencies except Apple's Network Extensions

**Comment by relikd** [comment]
here you go [proj.zip](https://github.com/relikd/AppCheck/files/5245713/proj.zip)

**Comment by MrBenFTW** [comment]
Thank you so much! this is exactly what I needed! I've been looking for something for over 3 months, kept coming across outdated projects in swift2 or they used NEKit. It became so much harder when apple added DNS over HTTPS support as everyone was talking about it.

Again, thanks so much :))

**Comment by MrBenFTW** [comment]
Is it possible to make it on demand? As in, it always connects

**Comment by relikd** [comment]
Haven't looked into it. I was searching for such a feature in the beginning,  funny enough I just found some resources right now by researching this. You would add this in `GlassVPN.createNewVPN()`, something like:
```
let rule = NEOnDemandRuleConnect()
rule.interfaceTypeMatch = .any
rule.probeURL = URL(string: "https://www.appchk.de")
mgr.onDemandRules = [rule]
// open var action: NEOnDemandRuleAction { get }
// open var dnsSearchDomainMatch: [String]?
// open var dnsServerAddressMatch: [String]?
// open var interfaceTypeMatch: NEOnDemandRuleInterfaceType
// open var ssidMatch: [String]?
// open var probeURL: URL?
```

see also apple documentation on [NEOnDemandRuleConnect](https://developer.apple.com/documentation/networkextension/neondemandruleconnect).

**Comment by MrBenFTW** [comment]
Not sure why but this doesn't seem to want to work over cellular. Nothing loads and it seems to time out. Yet over wifi it is working fine.

Checked settings > mobile data but the app is not there so it hasn't been disabled, with the app disconnected I can access sites via cellular so it is deffo an issue with Network extensions.

Do I have to enable something in the app to allow it access to mobile data?

**Comment by relikd** [comment]
This does not occur in AppCheck so I can't say for sure. A couple things I would try:

1. Disable and re-enable the Extension while on cellular. Maybe it gets a hiccup during switching? 
2. Try a different DNS resolver, maybe there is an issue with your own? Does it work with 1.1.1.1 or 8.8.8.8?
3. Try disabling the DNS altogether. Though I am not sure if the extension will work at all this way.
4. Is the bug related to the on-demand rule? Have you enabled those already? If so, try disabling it and see if it works.

But to answer your question, no. There is no option that needs to be enabled or similar. I think it should work right away.

**Comment by MrBenFTW** [comment]
Just tried it with the project you sent before since it doesn't have any changes or the on demand thingy, works over wifi but not cellular 🤔 maybe a bug in iOS 14?

**Comment by relikd** [comment]
Nah, AppCheck works fine on iOS 14. Must be related to the DNS settings (AppCheck does not use it at all) or your custom configuration. What does http://www.whatsmydnsserver.com/ say when you're on WiFi? Does it show your custom DNS or your home network resolver? And does the console print some useful information? You can filter for `[VPN.`.  
(Side note: If you try to debug the extension, don't use `print()` statements, they are not logged to the console. Use `NSLog` or `DDLog` instead)

**Comment by relikd** [comment]
I just tried this on iOS 13 and iOS 14, with WiFi and with cellular, all four combinations work fine. But I got an other idea why it might fail for you. You could try and add an IPv6 DNS server to that list and see if it helps.

**Comment by MrBenFTW** [comment]
Hi, randomly fixed itself 3 days ago. Not sure what fixed it as I've made quite a lot of changes but it works over cellular now ;)

if you are interested, the project is here: [Aesther](https://aesther.net) it is free, no ads etc.

**Comment by MrBenFTW** [comment]
How do I delete and re-save or overwrite the configuration? I want to be able to change the IP and also disable/enable on demand through my app, but it doesn't seem to save the new config in settings

**Comment by relikd** [comment]
The DNS IP is not stored in the configuration. It is set whenever the VPN starts. So you'll need to stop and restart the VPN with the new IP address. The communication between app and network extension is done either through `handleAppMessage` (while VPN is running) or by writing to the group bundle `UserDefaults`.

**Comment by MrBenFTW** [comment]
Oh, I know, I've already done all that :) But some stuff doesn't save. The DNS ip changes but when I set `mgr.isOnDemandEnabled = true` to` mgr.isOnDemandEnabled = false` it does not change the on demand rule. I have tried re launching the app, reconnecting etc but to no avail

**Comment by relikd** [comment]
Ahh, thats right, onDemand is set on the VPN setting, not the network extension. Yes, that part is a bit tricky. 
In [`GlassVPN.init()`][1] you will need to load the previous manager and delete it. I forgot what the function was but you'll find it.
Or you can also try and run the update func, see if that changes it.
```swift
self.updateVPN({ .. make your changes .. }) { NSLog("success") }
```
Other than that, I think `createNewVPN()` is only run once to init the system VPN. That is, until the user deletes the VPN in settings (Settings > General > VPN), then it will run again.

[1]: https://github.com/relikd/AppCheck/blob/main/main/GlassVPN.swift#L17-L29

**Comment by MrBenFTW** [comment]
Is it possible you could send me the one where you specify the DNS blocking in app with json? I am having issues still and I believe it is to do with caching. I'm trying to make an exclusive version of the homepage for aesther users who are connected to the dns server but ios is being weird about it. 

If you toggle airplane mode off and on then it will use the exclusive DNS server but after a few mins / if you exit safari, it will revert back to system dns

**Comment by relikd** [comment]
Could that be related to the on-demand setting or do you use the always-on VPN? Also, I don't understand your first question. Do you mean the project which uses NEKit? And what json are you referring to? I do not use json anywhere.

**Comment by MrBenFTW** [comment]
> What is your requirement exactly? Do you need to send the data to your DNS server or can this also be handled by the app solely?
> E.g. does your server contain complex logic or is it a simple domain list, that you want to block? In the latter case the app could contain that list and block outgoing traffic without ever contacting your server. This does not even need to be a static list. You could have an additional sync button inside your app that will update the blocking list from your server.

I was referring to the `can this also be handled by the app ` option, where the app will block it locally, no need for server. Based on remote list :) 

Also I don't think it is related to on demand as I've tried with it on and off

**Comment by relikd** [comment]
[DNS-with-NEKit.zip](https://github.com/relikd/AppCheck/files/5471995/DNS-with-NEKit.zip)

**Comment by relikd** [comment]
You can refer to the current AppCheck app for how I block requests. In particular if the list is very long, you'll want to use the binary tree lookup to match domains.
